Add domains list card UI

mkevins opened the pull request:
Fixes #19074

### Description

This PR adds composables for the domain card to be displayed in the domain list screen.

Note: In order to utilize the data classes from the API work, this has been pre-merged with https://github.com/wordpress-mobile/WordPress-Android/pull/19360. Some extra commits may appear below, until that PR is merged with a more recent version of `trunk`.

To test:

Check the designs via the Android Studio preview view:

| Preview |  Light Mode | Dark Mode |
|-|-|-|
| Status Rows |  <img width="526" alt="Screenshot 2023-10-13 at 4 59 29 PM" src="https://github.com/wordpress-mobile/WordPress-Android/assets/8507675/1b1793a8-397d-46ff-ba05-3a396f7f003e"> | <img width="531" alt="Screenshot 2023-10-13 at 4 59 47 PM" src="https://github.com/wordpress-mobile/WordPress-Android/assets/8507675/e7890328-5264-4a1b-a7ab-6bbaf7646d6e"> |
| Cards | ![image](https://github.com/wordpress-mobile/WordPress-Android/assets/8507675/0879d9bc-c8c4-4135-88a4-853b2849ab53) | ![image](https://github.com/wordpress-mobile/WordPress-Android/assets/8507675/693583ea-b513-4e78-b068-048146eb5bd5) |






## Regression Notes
1. Potential unintended areas of impact
N/A (The composables are not yet in use, except for in previews, i.e. no app usage yet).

2. What I did to test those areas of impact (or what existing automated tests I relied on)
N/A

3. What automated tests I added (or what prevented me from doing so)
N/A

PR submission checklist:

- [ ] I have completed the Regression Notes.
- [ ] I have considered adding accessibility improvements for my changes.
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:

- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] Talkback.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] Large and small screen sizes. (Tablet and smaller phones)
- [ ] Multi-tasking: Split screen and Pop-up view. (Android 10 or higher)

mkevins committed:
Add some studio color palette oranges to app colors

We'll need these in the card designs for the status row.

mkevins committed:
Add a material 3 theme with extra semantic colors

This is needed because the existing `AppTheme` does not support material
3. For now, this is added as specific theme for these designs, but it is
written with the possibility of hoisting it later for more general use
within the app.

mkevins committed:
Add a ui state class for the domain card

This can later be moved to the view model file when it is implemented.
Also, some of the details (e.g. status enum) may require changes /
refinements as the back-end is wired up.

mkevins committed:
Implement the domain status row composable

mkevins committed:
Implement domain card ui

Note: this does not yet include any click logic, but this can be handled
later, when there is something to wire it up to.

justtwago commented on WordPress/src/main/java/org/wordpress/android/ui/domains/management/DomainStatusRow.kt:
```diff
@@ -0,0 +1,90 @@
+    }
+}
+
+private val LocalDate.mediumFormat
+    get() = format(DateTimeFormatter.ofLocalizedDate(FormatStyle.MEDIUM))
+
+
+class PreviewStatusProvider: PreviewParameterProvider<DomainStatus> {
+    override val values
+        get() = DomainStatus.values().asSequence()
+}
```
What a nice way to add params to previews! 😮 👏

mkevins replied:
I was very pleased with this as well while coding it. 😄  Especially since the status enums are not yet finalized, afaik, so any changes needed there will automatically propagate to these previews.

justtwago commented on WordPress/src/main/java/org/wordpress/android/ui/domains/management/Theme.kt:
```diff
@@ -0,0 +1,94 @@
+        extraPaletteLight
+    }
+    CompositionLocalProvider (localColors provides extraColors) {
+        MaterialTheme(
+            colorScheme = if (isDarkTheme) {
+                darkColorScheme(
+                    error = AppColor.Red30,
+                )
+            } else {
+                lightColorScheme(
+                    error = AppColor.Red50,
+                )
+            },
+            content = content
+        )
+    }
```
I'm curious, why isn't this theme based on `AppTheme` from `org.wordpress.android.ui.compose.theme`? 

Btw, is it possible to reuse the existing `AppTheme` theme for the entire domain management feature instead of creating an extra `M3Theme`? We could expand the `JpLightColorPalette` and `JpDarkColorPalette` with the extra colors: success and warning.

justtwago replied:
In the future perspective, we could also extract the Jetpack Compose setup, such as theming to a separate module.

mkevins replied:
Good question. I forgot to mention that I left some info in the commit comments (e.g. this one is explained in [953cdc0](https://github.com/wordpress-mobile/WordPress-Android/pull/19269/commits/953cdc0992334ec6cdef9d70db456f441954186d)):

> This is needed because the existing `AppTheme` does not support material
> 3. For now, this is added as specific theme for these designs, but it is
> written with the possibility of hoisting it later for more general use
> within the app.

Does this make sense to you?

justtwago replied:
Yeah, totally! Thank you for the clarification!

mkevins committed:
Merge branch 'trunk' into issue/19074-domains-list-card-ui

mkevins committed:
Merge branch 'trunk' into issue/19074-domains-list-card-ui

mkevins committed:
Extract domain management status strings to resources

mkevins committed:
Add animated "ghost" strip for pending domain information

Also, this adjusts some of usages of semantic colors from the theme to
match the designs.

mkevins committed:
Merge branch 'trunk' into issue/19074-domains-list-card-ui

mkevins committed:
Merge remote-tracking branch 'origin/issue/19165-domains-list-endpoint' into issue/19074-domains-list-card-ui

mkevins committed:
Remove the domain status notice logic and UI

This is being removed as descoped for now.

mkevins committed:
Add gray 50 to app colors

mkevins committed:
Add some more extra colors

This commit also re-orders the definitions to satisfy lint.

mkevins committed:
Use domain status enum from FluxC

mkevins committed:
Update domain status row for new ui state types

This also makes expiry optional (via a pending ghost strip).

mkevins committed:
Update ghost strip colors

mkevins committed:
Update domain card for new ui state types

This also makes all text fields optional (via ghost strip).

mkevins committed:
Fix detekt issue

antonis committed:
Merge branch 'trunk' into issue/19074-domains-list-card-ui

antonis commented on WordPress/src/main/java/org/wordpress/android/ui/domains/management/Theme.kt:
```diff
@@ -0,0 +1,111 @@
+private val localColors = staticCompositionLocalOf { extraPaletteLight }
+
+@Composable
+fun M3Theme(
```
[TIL](https://developer.android.com/jetpack/compose/designsystems/material3#material-theming) :)

antonis commented on WordPress/src/main/res/values/strings.xml:
```diff
@@ -2698,6 +2698,18 @@
     <string name="domains_search_for_a_domain_button">Search for a domain</string>
 
+    <!-- Domain Management -->
+    <string name="domain_management_status_complete_setup">Complete Setup</string>
```
My understanding is that those are not used at this point and it is pending to decide if we will use the [localised server status string](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/trunk/fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpcom/site/AllDomainsResponse.kt#L53) instead. Is that correct?

mkevins replied:
That's correct, afaiu. The only exception is the error case where a status is somehow `null`. In that case, we can use an app-translated "Error" string to convey some error state to the user. Resolved in  https://github.com/wordpress-mobile/WordPress-Android/pull/19269/commits/bfe59b3f0c85424d77dde802d6687a957a5777b1 and https://github.com/wordpress-mobile/WordPress-Android/pull/19269/commits/8b0de936127868384306438116cf5b25f3617a7f.

antonis replied:
Thank you for the change @mkevins 🙇

antonis reviewed (approved):
Great work @mkevins 👍 
The code LGTM and the UI worked as expected both in preview and on device in all configurations 🎉 
I'll restrain from merging in case @justtwago has more feedback especially on the Compose side ;)

justtwago commented on WordPress/src/main/java/org/wordpress/android/ui/domains/management/DomainStatusRow.kt:
```diff
@@ -18,70 +18,107 @@ import androidx.compose.runtime.Composable
+            BulletPoint(color = status.indicatorColor)
+            Spacer(modifier = Modifier.width(8.dp))
+            Text(
+                text = status.status ?: "Unknown",
```
Should `"Unknown"` be also moved to `strings.xml`?

mkevins replied:
Good catch! I left this here as a sort of "placeholder" until we figure out what to do about the status strings. The current plan is to have them translated on the backend, but in case the response is empty or null for this field. In this case, the string should be "Error", so this has been resolved w/ https://github.com/wordpress-mobile/WordPress-Android/pull/19269/commits/bfe59b3f0c85424d77dde802d6687a957a5777b1 and https://github.com/wordpress-mobile/WordPress-Android/pull/19269/commits/8b0de936127868384306438116cf5b25f3617a7f.

justtwago commented on WordPress/src/main/java/org/wordpress/android/ui/domains/management/DomainStatusRow.kt:
```diff
@@ -18,70 +18,107 @@ import androidx.compose.runtime.Composable
+        Spacer(modifier = Modifier.weight(1f))
+        expiry?.also { expiry ->
+            Text(
+                text = "Expires ${expiry.mediumFormat}",
```
And same here. If it's not temporary, I'd put it to `strings.xml`

mkevins replied:
Good idea. Resolved w/ https://github.com/wordpress-mobile/WordPress-Android/pull/19269/commits/f3a00c199ee8f07c9f6ff764faecfbbacb41d79c#diff-6bbab92723bb83a1300e86d496d62b9475a6a1e8b5a5feca27721edfbb09943fR2702.

mkevins committed:
Create fallback error string for domain status

Generally, the domain status string comes from the back-end, but in
cases where it is `null`, we need to display an app-translated error
string.

mkevins committed:
Treat UNKNOWN and null statusTypes as ERROR

mkevins committed:
Add a ui state for the status row

mkevins committed:
Refactor domain status row to use status row ui state

mkevins committed:
Refactor DomainCardUiState as a sealed class

This allows for specifying the initial state separately from implying
that from null fields. This is useful for hiding optional strings
(to avoid displaying a pending animation that will never load).

mkevins committed:
Refactor domain list card to use the sealed class ui state

mkevins committed:
Remove unused strings

mkevins commented:
Thank you Antonis and Artyom for reviewing this! I have pushed some commits to address the remaining items. I've also done some small refactoring to tidy things, and to allow for an initial state (ghost strips) as distinct from hiding optional strings entirely.

This is ready for another look.

antonis commented on WordPress/src/main/java/org/wordpress/android/ui/domains/management/DomainsListCard.kt:
```diff
@@ -0,0 +1,174 @@
+                }
+            }
+            Spacer(modifier = Modifier.width(16.dp))
+            Icon(
```
This is a suggestion, not a change request, and can be handled on a follow up PR. Would it make sense to hide the arrow Icon while loading in the Initial ui state? cc @paulforgione


https://github.com/wordpress-mobile/WordPress-Android/assets/304044/eb15a8db-8a09-4ed7-9517-9d41a327c7f6

mkevins replied:
Good point. I'll defer to Paul, but we can address in one of the followup PRs if needed.

antonis reviewed (approved):
Thank you for the changes @mkevins 🙇 
The code LGTM and works as expected. I've tested with the preview functionality and via https://github.com/wordpress-mobile/WordPress-Android/pull/19366 and everything worked for me on a Pixel 5 (Android 14) 🎉 

I left a minor suggestion here https://github.com/wordpress-mobile/WordPress-Android/pull/19269/files#r1363361955

mkevins commented:
Thank you for reviewing and testing this out!
